### ISSUE TITLE
Converting 2D self-attention layer into 3D

### ISSUE THREAD

**Issue by michalSkoury** [MV-Dream/MVDream#6]
Hey great work!
Small question when you mentioned: " We can convert original 2D self-attention layer into 3D by **connecting** different views in the self-attention layer". Did you mean you concatenated the images to get not only self attention on a single view but also cross attention between different views?

**Comment by lorenmt** [comment]
I was wondering the same thing. But some simple experiments show that does not work well, mainly I guess the cross attention between other views pushes the representations out-of-distribution?

**Comment by seasonSH** [comment]
Yes. And you should observe that the model generates similar images for the 4 views (even without training) once you convert 2D self-attention to 3D self+crossframe attention.

**Comment by Friedrich-M** [comment]
I was also wondering how you connect different views in the self-attention layer. The following code is the implementation of the temporal attention module.

<img width="587" alt="image" src="https://github.com/MV-Dream/MVDream/assets/85838942/1057ff90-97b9-4f53-af71-0172f0040b47">

Did you rearrange the "hidden states" from the size of "(b f) d c" to "b (f d) c" to connect the different views?

**Comment by seasonSH** [comment]
@Friedrich-M 
No. I assume your d=HxW, which is the image size. Then it should be like this:
``` python3
hidden_states = rearrange(hidden_states, "(b f) d c -> b (f d) c", f=sequence_length)
hidden_states = self.attn1(self.norm1(hidden_states)) + hidden_states
hidden_states = rearrange(hidden_states, "b (f d) c -> (b f) d c", f=sequence_length)
```
where attn1 is the original 2D self-attention in latent diffusion that we reuse.

**Comment by Friedrich-M** [comment]
Thanks a lot for your response! I fully understand it.
